user_sync/connector/directory_ldap.py

```diff
@@ -269,35 +272,55 @@ def iter_users(self, users_filter):
                 if (last_attribute_name != None):
                     self.logger.warn('No email attribute: %s for dn: %s', last_attribute_name, dn)
                 continue
+
+            source_attributes = {}
             
             user = user_sync.connector.helper.create_blank_user()
             user['email'] = email
+            source_attributes['email'] = email
                 
             username, last_attribute_name = self.user_username_formatter.generate_value(record)
             if (username == None and last_attribute_name != None):
                 self.logger.info('No username attribute: %s for dn: %s', last_attribute_name, dn)    
             user['username'] = username if username != None else email
+            if (username != None):
```

Comment:
Why is this conditionalized?  Seems like you should be setting the `source_attributes['username']` to be `user['username']` so that it always has a value.  Not setting it at all is likely to lead to assertions when it's accessed in the user extension code.